File: v4/rust/iprange-livedb/src/recovery/range_components.rs
//! One ordered overlap-component pass for recovered ranges.

use crate::cancellation::CancellationToken;
use crate::error::{Error, Result};
use crate::range_tree::Record;

use super::direct_output::DirectKey;
use super::range_build::RangeOutput;

pub(super) trait Policy<K: DirectKey> {
    type Resolved: Copy;

    fn resolve(&mut self, record: Record<K>) -> Result<Option<Self::Resolved>>;
    fn accept(&mut self, record: Record<K>, resolved: Option<Self::Resolved>) -> Result<()>;
    fn reject_overlap(&mut self, count: u64, from: K, to: K) -> Result<()>;
    fn finish(&mut self) -> Result<()>;
}

pub(super) struct Components<'a, K: DirectKey, P: Policy<K>> {
    cancellation: &'a CancellationToken,
    policy: P,
    component: Option<Component<K, P::Resolved>>,
}

#[derive(Clone, Copy)]
struct Component<K, R> {
    first: Record<K>,
    resolved: Option<R>,
    maximum_to: K,
    count: u64,
}

impl<'a, K: DirectKey, P: Policy<K>> Components<'a, K, P> {
    pub(super) fn new(cancellation: &'a CancellationToken, policy: P) -> Self {
        Self {
            cancellation,
            policy,
            component: None,
        }
    }

    fn push(&mut self, record: Record<K>) -> Result<()> {
        self.cancellation.check()?;
        let resolved = self.policy.resolve(record)?;
        let Some(mut component) = self.component.take() else {
            self.component = Some(Component::new(record, resolved));
            return Ok(());
        };
        if record.from < component.first.from {
            return Err(Error::RecoveryCandidateChanged);
        }
        if record.from <= component.maximum_to {
            component.maximum_to = component.maximum_to.max(record.to);
            component.count = component
                .count
                .checked_add(1)
                .ok_or(Error::ArithmeticOverflow("recovery overlap component"))?;
            self.component = Some(component);
            return Ok(());
        }
        self.finish_component(component)?;
        self.component = Some(Component::new(record, resolved));
        Ok(())
    }

    fn finish(&mut self) -> Result<()> {
        if let Some(component) = self.component.take() {
            self.finish_component(component)?;
        }
        self.policy.finish()
    }

    fn finish_component(&mut self, component: Component<K, P::Resolved>) -> Result<()> {
        if component.count != 1 {
            self.policy
                .reject_overlap(component.count, component.first.from, component.maximum_to)
        } else {
            self.policy.accept(component.first, component.resolved)
        }
    }
}

impl<K: DirectKey, P: Policy<K>> RangeOutput<K> for Components<'_, K, P> {
    fn push(&mut self, record: Record<K>) -> Result<()> {
        Components::push(self, record)
    }

    fn finish(&mut self) -> Result<()> {
        Components::finish(self)
    }
}

impl<K: Copy, R: Copy> Component<K, R> {
    fn new(first: Record<K>, resolved: Option<R>) -> Self {
        Self {
            first,
            resolved,
            maximum_to: first.to,
            count: 1,
        }
    }
}

File: v4/go/internal/recovery/range_components.go
package recovery

// One ordered overlap-component pass for recovered ranges (Rust
// recovery/range_components.rs): adjacent-disjoint readable records
// stream into the policy, overlapping records accumulate one component
// that is rejected as a whole with its fence, and an order regression
// is the candidate-changed class.

import "github.com/firehol/iprange/v4/go/internal/live"

// rangePolicy consumes the resolved components of one overlap pass
// (Rust Policy).
type rangePolicy interface {
	resolve(record rangeRecord) (any, error)
	accept(record rangeRecord, resolved any) error
	rejectOverlap(count uint64, from, to rangeKey) error
	finish() error
}

// component is one accumulated overlap component (Rust Component): the
// first record, the resolved token, the maximum to-key, and the record
// count.
type component struct {
	first     rangeRecord
	resolved  any
	maximumTo rangeKey
	count     uint64
}

// components is the ordered overlap-component pass (Rust Components).
type components struct {
	check     func() error
	codec     rangeCodec
	policy    rangePolicy
	component *component
}

// push accepts one record into the pass (Rust Components::push: the
// checkpoint, the policy resolution, the order proof, and the overlap
// accumulation or component finish).
func (c *components) push(record rangeRecord) error {
	if err := live.Checkpoint(c.check); err != nil {
		return err
	}
	resolved, err := c.policy.resolve(record)
	if err != nil {
		return err
	}
	if c.component == nil {
		c.component = &component{first: record, resolved: resolved, maximumTo: record.to, count: 1}
		return nil
	}
	current := *c.component
	c.component = nil
	if c.codec.lessKey(record.from, current.first.from) {
		return candidateChangedError()
	}
	if !c.codec.lessKey(current.maximumTo, record.from) {
		// record.from <= maximum_to: the record belongs to the open
		// component.
		if c.codec.lessKey(current.maximumTo, record.to) {
			current.maximumTo = record.to
		}
		next := current.count + 1
		if next == 0 {
			return overflowError("recovery overlap component")
		}
		current.count = next
		c.component = &current
		return nil
	}
	if err := c.finishComponent(current); err != nil {
		return err
	}
	c.component = &component{first: record, resolved: resolved, maximumTo: record.to, count: 1}
	return nil
}

// finish closes the open component and the policy (Rust
// Components::finish).
func (c *components) finish() error {
	if c.component != nil {
		current := *c.component
		c.component = nil
		if err := c.finishComponent(current); err != nil {
			return err
		}
	}
	return c.policy.finish()
}

// finishComponent resolves one closed component (Rust
// Components::finish_component: one record is accepted, a multi-record
// component is rejected as a whole).
func (c *components) finishComponent(current component) error {
	if current.count != 1 {
		return c.policy.rejectOverlap(current.count, current.first.from, current.maximumTo)
	}
	return c.policy.accept(current.first, current.resolved)
}
